test/Microsoft.AspNet.Mvc.Xml.Test/XmlAssert.cs

```diff
@@ -20,30 +23,65 @@ public static class XmlAssert
         /// <param name="actualXml">Actual xml string.</param>
         public static void Equal(string expectedXml, string actualXml)
         {
-            var sortedExpectedXDoc = SortAttributes(XDocument.Parse(expectedXml));
-            var sortedActualXDoc = SortAttributes(XDocument.Parse(actualXml));
+            var sortedExpectedXDocument = SortAttributes(XDocument.Parse(expectedXml));
+            var sortedActualXDocument = SortAttributes(XDocument.Parse(actualXml));
 
-            bool areEqual = XNode.DeepEquals(sortedExpectedXDoc, sortedActualXDoc);
+            // Since XNode's DeepEquals does not check for presence of xml declaration,
+            // check it explicitly
+            bool areEqual = EqualDeclarations(sortedExpectedXDocument.Declaration, sortedActualXDocument.Declaration);
+
+            areEqual = areEqual && XNode.DeepEquals(sortedExpectedXDocument, sortedActualXDocument);
 
             if (!areEqual)
             {
-                throw new EqualException(sortedExpectedXDoc, sortedActualXDoc);
+                throw new EqualException(GetRawXml(sortedExpectedXDocument), GetRawXml(sortedActualXDocument));
             }
         }
 
-        private static XDocument SortAttributes(XDocument doc)
+        private static bool EqualDeclarations(XDeclaration expected, XDeclaration actual)
+        {
+            if (expected == null && actual == null)
+            {
+                return true;
+            }
+
+            if (expected == null || actual == null)
+            {
+                return false;
+            }
+
+            return string.Equals(expected.Version, actual.Version, StringComparison.OrdinalIgnoreCase)
+                && string.Equals(expected.Encoding, actual.Encoding, StringComparison.OrdinalIgnoreCase);
+        }
+
+        private static XDocument SortAttributes(XDocument document)
         {
             return new XDocument(
-                    doc.Declaration,
-                    SortAttributes(doc.Root));
+                    document.Declaration,
+                    document.Nodes().Where(node => node.NodeType != XmlNodeType.Element),
+                    SortAttributes(document.Root));
         }
 
         private static XElement SortAttributes(XElement element)
         {
             return new XElement(
                     element.Name,
+                    element.Nodes().Where(node => node.NodeType != XmlNodeType.Element),
                     element.Attributes().OrderBy(a => a.Name.ToString()),
                     element.Elements().Select(child => SortAttributes(child)));
```

Comment:
the `NodeType != Element` ... `element.Elements()` pattern will reorder the nodes oddly.  won't affect the equality checks but will make it harder to read the `Message` of the `EqualException` when something is different.  e.g. compare `<doc>The <b>brown</b> dog</doc>` with `<doc>The <b>white</b> dog</doc>` and see errors about `<doc>The dog <b>brown</b></doc>` not matching `<doc>The dog <b>white</b></doc>`